### ISSUE TITLE
Name multicat columns

### ISSUE THREAD

**Issue by lmmx** [lmmx/multicorr#1]
Context: ([Aiming for something more like R's factors than pandas categories due to the latter's omission of null handling](https://pandas.pydata.org/pandas-docs/stable/user_guide/categorical.html#differences-to-r-s-factor))

I can't experiment with shorter combinations of categories until I can identify the factor (not just the level) in the col name, and my existing approach just involved `apply`ing `"_".join` to the series.

Instead we can name the 'multicategory' attaching the category name to each component level (and vice versa)

https://github.com/lmmx/multicorr/blob/b81e06807248db3160f20a64a010ff1382c66e52/corr_analysis_multicat_dummy_form.py#L19-L26

I want to use this, in the same way I'm using the simple level names, with dummy vars

Originally simply on the data

https://github.com/lmmx/multicorr/blob/b81e06807248db3160f20a64a010ff1382c66e52/corr_analysis_multicat_dummies.py#L15

but now I want to apply it to just the `null_combos` (no leverage in multifactors without nulls)

https://github.com/lmmx/multicorr/blob/b81e06807248db3160f20a64a010ff1382c66e52/corr_analysis_multicat_dummy_form.py#L29

So we get a single series (row) and the default behaviour is to `_` conjoin the column name (factor) and its value (level):

```py
>>> null_combos
  c1 c2 c3
3  B  B  C
>>> pd.get_dummies(null_combos[cat_cols].astype(str))
   c1_B  c2_B  c3_C
3     1     1     1
```

We want to **override the naming** but otherwise not interfere with this, i.e. the goal is

```py
   c1_B:c2_B:c3_C
3               1
```

Put another way, pandas [only supports 1D](https://github.com/pandas-dev/pandas/blob/v1.5.3/pandas/core/reshape/encoding.py#L25-L213) for its one hot encoding, and we essentially want 2D, so will need to transform the data to make 1 hot features from it (equivalent to squeezing, and in fact this may be the best way to do so).

**Comment by lmmx** [comment]
So we get a single series (row)

```py
>>> null_combos
  c1 c2 c3
3  B  B  C
```

and the default behaviour is to `_` conjoin the column name (factor) and its value (level):

```py
>>> pd.get_dummies(null_combos[cat_cols].astype(str))
   c1_B  c2_B  c3_C
3     1     1     1
```

We want to **override the naming** but otherwise not interfere with this

**Comment by lmmx** [comment]
```py
s = null_combos.loc[3]
sdf = pd.Series({name_multicat(s): 1}).to_frame().T
```

gives

```py
>>> sdf
   c1_B:c2_B:c3_C
0               1
```

**Comment by lmmx** [comment]
This seems to work (I added another matching row to the toy dataset to help):

```py
>>> null_combos.apply(lambda s: pd.Series({name_multicat(s): 1}), axis=1).fillna(0).astype(int)
   c1_B:c2_B:c3_C  c1_D:c2_D:c3_C
3               1               0
8               0               1
```

To use these as dummies they'd need to be expanded back out to the length of the full dataset.

**Comment by lmmx** [comment]
Again we can pop out a row at a time (beginning with the first row we have as a positive result, with a `NaN`).

```py
dummy_variables = null_combos.apply(lambda s: pd.Series({name_multicat(s): 1}), axis=1).fillna(0).astype(int)
s = data[original_cols].loc[3]
row_name = name_multicat(s[cat_cols].astype(str)
row_name in dummy_variables.columns
```

gives a bool, or we can do this in the pandas API (maybe more vectorised? IDK) by reversing it

```py
dummy_variables.columns.str.fullmatch(row_name).any()
```

**Comment by lmmx** [comment]
So now we have `dummy_2d` which creates the dummy cols

https://github.com/lmmx/multicorr/blob/1610aa55ee3a9af48cda91eb039c0d1db9d659f7/corr_analysis_multicat_dummy_form.py#L19-L30


https://github.com/lmmx/multicorr/blob/1610aa55ee3a9af48cda91eb039c0d1db9d659f7/corr_analysis_multicat_dummy_form.py#L36-L37

Now we need to create `dummy_check` which can apply across columns to indicate the dummy vars (multiple in anticipation of future not-all-columns matching) to activate at a given row

This works:

```py
pd.Series(dict(zip(dummy_cols, dummy_cols.str.fullmatch(name_multicat(s.astype(str))).astype(int))))
```

Though I worry about performance of a dict zip (not clear how to retain the dummy_cols as a series: keep index and insert maybe?)

```py
>>> pd.Series(dummy_cols.str.fullmatch(name_multicat(s.astype(str))).astype(int), index=dummy_cols)
c1_B:c2_B:c3_C    0
c1_D:c2_D:c3_C    0
dtype: int64
```

**Comment by lmmx** [comment]
Successfully implemented 2D dummies fb12651afb80e32f701ef7a59397b0d4581a3d67

```py
>>> data_with_dummies                                                          
   id c1 c2 c3     v  null_target_column  c1_B:c2_B:c3_C  c1_D:c2_D:c3_C
0   1  A  A  C  10.0                   0               0               0
1   2  B  A  C  20.0                   0               0               0
2   3  A  B  C  30.0                   0               0               0
3   4  B  B  C   NaN                   1               1               0
4   5  A  A  C  50.0                   0               0               0
5   6  B  A  C  60.0                   0               0               0
6   7  A  B  C  70.0                   0               0               0
7   8  B  B  C   NaN                   1               1               0
8   9  D  D  C   NaN                   1               0               1
```

The next step is to make the assertion pass again, which will require whittling down the extra unused degree of freedom (the extraneous `c3`)

**Comment by lmmx** [comment]
The next step is to make the assertion pass again, which will require whittling down the extra unused degree of freedom (the extraneous `c3`)
